### PR TITLE
[cuDNN] Add CuDNNExecutable to encapsulate cuDNN execution plans

### PR DESCRIPTION
**Pull request by ezhulenev**

### CODE REVIEW

File: runtime/src/openxla/runtime/nvgpu/cudnn_api.cpp [hunk 1]
```diff
@@ -93,16 +94,30 @@ CuDNNOperationGraph::~CuDNNOperationGraph() {
   graph_.reset();
 }
 
-const cudnn_frontend::OperationGraph& CuDNNOperationGraph::graph() const {
-  return *graph_;
+cudnn_frontend::OperationGraph& CuDNNOperationGraph::graph() { return *graph_; }
+
+//===----------------------------------------------------------------------===//
+// CuDNNExecutable.
```
Comment (chsigg) [inline runtime/src/openxla/runtime/nvgpu/cudnn_api.cpp:100]: I think you wanted to have those without the `.`?

File: runtime/src/openxla/runtime/nvgpu/cudnn_api.cpp [hunk 2]
```diff
@@ -203,13 +320,97 @@ StatusOr<vm::ref<CuDNNOperationGraph>> CreateOperationGraph(
       new CuDNNOperationGraph(syms, std::move(graph)));
 }
 
+//===----------------------------------------------------------------------===//
+// CreateExecutable.
```
Comment (chsigg) [inline runtime/src/openxla/runtime/nvgpu/cudnn_api.cpp:324]: Without `.`?

File: runtime/src/openxla/runtime/nvgpu/cudnn_api.h [hunk 3]
```diff
@@ -106,13 +108,34 @@ class CuDNNOperationGraph : public iree::vm::RefObject<CuDNNOperationGraph> {
...
-  const cudnn_frontend::OperationGraph& graph() const;
+  cudnn_frontend::OperationGraph& graph();
 
  private:
   openxla_cudnn_dynamic_symbols_t* syms_;
   std::optional<cudnn_frontend::OperationGraph> graph_;
 };
 
+//===----------------------------------------------------------------------===//
+// CuDNN executable.
```
Comment (chsigg) [inline runtime/src/openxla/runtime/nvgpu/cudnn_api.h:119]: Without `.`?

**Review by chsigg — APPROVED** [review]
I assume we will always create a single executable from every graph. Is there any benefit of separating `!cudnn.graph` and `!cudnn.executable` beyond being truthful to the cuDNN API? 
(... which would be enough of a motivation, I'm just asking).

Do you intend to expose the heuristics as attributes to `cudnn.graph`, or simply add them during lowering to runtime?

File: runtime/src/openxla/runtime/nvgpu/cudnn_api.cpp (hunk 1 above)
Comment (ezhulenev) [reply runtime/src/openxla/runtime/nvgpu/cudnn_api.cpp:100]: Yeah, adding a little bit more consistency with IREE style

**Comment by ezhulenev** [comment]
And if no explicit auto-tuning or plan selection is done in the IR, then executable itself will decide what plan to run (it might also do run and measure tuning at run time)

**Comment by chsigg** [comment]
Thanks, that makes sense.

>> Do you intend to expose the heuristics as attributes

I was thinking that auto tuning might happen aot at the cudnn dialect level and you would stick the result to the cudnn.graph op.

**Pull request merged**